### PR TITLE
#188 use JSON to serialize complex objects over to JS

### PR DESCRIPTION
**Pull request by jfrantzius**
instead of using Native\* API from org.mozilla package namespace

### CODE REVIEW

File: handlebars/src/main/java/com/github/jknack/handlebars/internal/js/RhinoHandlebars.java [hunk 1]
```diff
@@ -246,6 +169,39 @@ public CharSequence apply(final Object context, final Options options) throws IO
...
+    if (object instanceof Number) {
+      return true;
+    }
+    if (object instanceof Boolean) {
+      return true;
+    }
+    if (object instanceof CharSequence || object instanceof Character) {
+      return true;
+    }
+    if (object instanceof Scriptable) {
```
Comment (jknack) [inline handlebars/src/main/java/com/github/jknack/handlebars/internal/js/RhinoHandlebars.java]: Should we make a java.util.Date a simple object?

File: handlebars/src/main/resources/helpers.rhino.js [hunk 2]
```diff
@@ -18,37 +18,76 @@ Handlebars.SafeString = com.github.jknack.handlebars.Handlebars.SafeString;
...
-    };
-  
-    Handlebars.helpers[name] = helper;
-    Handlebars_java.registerHelper(name, fn);
-  };
+/**
+ * Wrap OptionsJs with JSON deserialization of hash and params,
+ * so they become accessible as native JS object and array, respectively. 
+ */
+function JsOptions(javaOptionsJs) {
```
Comment (jknack) [inline handlebars/src/main/resources/helpers.rhino.js]: move this function inside the anom fn (line 41)

**Comment by jknack** [comment]
looks good!

File: handlebars/src/main/java/com/github/jknack/handlebars/internal/js/RhinoHandlebars.java (hunk 1 above)
Comment (jfrantzius) [inline handlebars/src/main/java/com/github/jknack/handlebars/internal/js/RhinoHandlebars.java]: Hm, I'd guess that GSON's serialization + deserialization results in something better than letting JS talk to the Java Date object directly, as that probably doesn't support [JS date object methods](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Date#Methods_2) . I didn't research it, though, whether it's really proper JS date object coming out of JSON serialization/deserialization with GSON.

Would be interesting to see what GSON assumes as timezone.

File: handlebars/src/main/resources/helpers.rhino.js (hunk 2 above)
Comment (jfrantzius) [inline handlebars/src/main/resources/helpers.rhino.js]: Sorry what's an anom fn? My Javascript isn't all that good :)

Comment (jknack) [inline handlebars/src/main/resources/helpers.rhino.js]: I mean this:

```js
(function () {
   function JsOptions(javaOptionsJs) {
   }
   ...
})();
```

The ```(function () {...})``` is a js anonymous function :)

**Comment by jfrantzius** [comment]
Hi @jknack ,
I'm now using Nashorn object translation by default on Java 8, which works with `Java version: 1.8.0_45, vendor: Oracle Corporation`. Current test failures on Java 8 are likely due to JDK version not being 1.8.0_45 on Travis.

For Java 7, an optional module / dependency exists, which is included with test scope when build is run on Java 7. I was going for Rhino bundled with Oracle Java 7, so no further Rhino dependency is required there, but the package `sun.org.mozilla.javascript.internal` doesn't exist on OpenJDK 7. 

Maybe it's better to go with a separate Rhino dependency for both Java 7 and Java 6 and then rely on the `org.mozilla.javascript.internal` package (without `sun` prefix). Not sure how we can distinguish builtin Rhino and Rhino from dependency at runtime, because they'd then both be present when building (and running) on Java 7.

**Comment by matteosilv** [comment]
Hi, may i ask why this pull request has not been merged into the master?

**Comment by jknack** [comment]
Code has been simplified a lot now we upgrade to java 8 and nashorn. See https://github.com/jknack/handlebars.java/commit/a980a50c7c0316f37e7a359989efb0ff1179f6ba

**Pull request closed without merging**